base/loading.jl

```diff
@@ -1779,7 +1784,8 @@ function __require(into::Module, mod::Symbol)
         end
         uuidkey, env = uuidkey_env
         if _track_dependencies[]
-            push!(_require_dependencies, (into, binpack(uuidkey), 0.0))
+            path = binpack(uuidkey)
+            push!(_require_dependencies, (into, path, UInt64(0), UInt32(0), replace_depot_path(path)))
```

Comment:
Why do we split `path` and `depot_alias`? 

I think we should just do: ` push!(_require_dependencies, (into, replace_depot_path(path), UInt64(0), UInt32(0)))`
and teach all consumsers how to handle `@depot` prefixes.

Reply:
Been trying to update consumers but I am stuck at `precompile.c:write_srctext`:
This is the only place in the C code where the `abspath` is needed. I tried to reconstruct it from the depot alias by calling back into `Base.replace_depot_alias` using something like
```diff
diff --git a/src/precompile.c b/src/precompile.c
index 51ddea37b1..83d0f289c0 100644
--- a/src/precompile.c
+++ b/src/precompile.c
@@ -36,6 +36,10 @@ void write_srctext(ios_t *f, jl_array_t *udeps, int64_t srctextpos) {
         //   char*: src text
         // At the end we write int32(0) as a terminal sentinel.
         size_t len = jl_array_len(udeps);
+        static jl_value_t *resolve_depot_func = NULL;
+        if (!resolve_depot_func)
+            resolve_depot_func = jl_get_global(jl_base_module, jl_symbol("resolve_depot_path"));
+        jl_value_t *resolve_depot_args[2] = {resolve_depot_func, NULL};
         ios_t srctext;
         for (size_t i = 0; i < len; i++) {
             jl_value_t *deptuple = jl_array_ptr_ref(udeps, i);
@@ -43,8 +47,10 @@ void write_srctext(ios_t *f, jl_array_t *udeps, int64_t srctextpos) {
             // Dependencies declared with `include_dependency` are excluded
             // because these may not be Julia code (and could be huge)
             if (depmod != (jl_value_t*)jl_main_module) {
-                jl_value_t *dep = jl_fieldref(deptuple, 1);  // file abspath
-                const char *depstr = jl_string_data(dep);
+                jl_value_t *depalias = jl_fieldref(deptuple, 1);  // file @depot alias
+                resolve_depot_args[1] = depalias;
+                jl_value_t *abspath = (jl_value_t*)jl_apply(resolve_depot_args, 2);
+                const char *depstr = jl_string_data(abspath);
                 if (!depstr[0])
                     continue;
                 ios_t *srctp = ios_file(&srctext, depstr, 1, 0, 0, 0);
@@ -53,7 +59,6 @@ void write_srctext(ios_t *f, jl_array_t *udeps, int64_t srctextpos) {
                               depstr);
                     continue;
                 }
-                jl_value_t *depalias = jl_fieldref(deptuple, 4);  // file @depot alias
                 size_t slen = jl_string_len(depalias);
                 write_int32(f, slen);
                 ios_write(f, jl_string_data(depalias), slen);

```

But this leaves me with error messages of the following form during the `Collecting and executing precompile statements` of building:
```julia
MethodError(f=Base.resolve_depot_path, args=("/tmp/jl_bRV80h/__PackagePrecompilationStatementModule/src/__PackagePrecompilationStatementModule.jl",), world=0x0000000000000001)
```

Any idea what's the problem here?
Am I issuing a wrong call?

[I know that the `replace_depot_path` method needs to be updated too, because atm it can throw, but I think that is not the issue here because locally I already made it to not throw.]

Reply:
So the method error says that this method doesn't exist in that world. 

You might need to do:

```
    jl_task_t *ct = jl_current_task;
    size_t last_age = ct->world_age;
    ct->world_age = jl_atomic_load_acquire(&jl_world_counter);
    jl_value_t *result = jl_apply(args, 6);
    ct->world_age = last_age;
```

Or something similar.

Reply:
Worked, thanks!

Reply:
I wonder if there is something to be gained from reverting these changes (e.g. going back to only forwarding the `abspath` to C) but insert the `@depot` tags very late and just before writing?
I didn't realized before that this would also be an option. What do you think?

Reply:
I think the current strategy is fine.